## 4G Signal Strength Mapping in UVA Engineering School
### WIoT Final Project Group 5

**Project Motivation**:
Long range signal transmission can have blind spots well within their proposed range–as we have seen for LoRa and BLE, where the theoretical range is a lot shorter or is significantly reduced due to the objects in the environment. LoRa has a proposed range of 3 miles in urban areas, but as shown in the results of the measurements in this class, it can have certain blind spots well within the 3 mile range, especially in underground settings such as basements. 4G/5G are important means of communication on Grounds, and they are relied upon for real time communication in case WiFi fails. Yet, it is sometimes reported that the 4G/5G coverage on Grounds is not ideal, especially in terms of underground environments. The Engineering School has a lot of buildings where the basements are environments where important experiments take place. Having robust and redundant communication is important for the safety of both the faculty, staff, students, and their work. Hence, we propose to measure the 4G/5G coverage in the E-School and around grounds. We will use various internet speed measurement apps on our phones to collect the data at specific geographic locations, indoors and outdoors. Using this data we can then create heatmaps for both indoor and outdoor settings, exploring the differences and enlightening the community to areas of low coverage.

**This python notebook is step 3 of our project where we map the collected signal strength data**

Data Mapping Plan:
1. map data by type of sinal strenght (upload, download, rssi), floors and their coordinates.
2. Collect stats, such as mean, std, and range by floor, building and graph bar chart

Library Setup and Data Import

In [75]:
import pandas as pd
import numpy as np
import plotly.express as px
print('Note: all ploly graphs are interactive, to gain further (quantitative, other data) for a data point, hover your cursr above a dot')

Note: all ploly graphs are interactive, to gain further (quantitative, other data) for a data point, hover your cursr above a dot


In [7]:
indoors = pd.read_csv("WIoT Final Project 4G_5G Data - Indoor.csv")
outdoors = pd.read_csv("WIoT Final Project 4G_5G Data - Outdoor.csv")

Data Exploration and Processing

In [4]:
print(outdoors.columns)
print(outdoors.info())

Index(['Location ID', 'Description', 'Lat', 'Long', '4G RSSI signal',
       '4G download rate', '4G upload rate'],
      dtype='object')
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22 entries, 0 to 21
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Location ID       22 non-null     object 
 1   Description       22 non-null     object 
 2   Lat               22 non-null     float64
 3   Long              22 non-null     float64
 4   4G RSSI signal    22 non-null     int64  
 5   4G download rate  22 non-null     float64
 6   4G upload rate    22 non-null     float64
dtypes: float64(4), int64(1), object(2)
memory usage: 1.3+ KB
None


In [5]:
print(indoors.columns)
print(indoors.info())

Index(['Location ID', 'Description', 'Floor', 'Lat', 'Long', '4G RSSI signal',
       '4G download rate', '4G upload rate'],
      dtype='object')
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79 entries, 0 to 78
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Location ID       79 non-null     object 
 1   Description       79 non-null     object 
 2   Floor             79 non-null     object 
 3   Lat               79 non-null     float64
 4   Long              79 non-null     float64
 5   4G RSSI signal    79 non-null     int64  
 6   4G download rate  79 non-null     float64
 7   4G upload rate    79 non-null     float64
dtypes: float64(4), int64(1), object(3)
memory usage: 5.1+ KB
None


# Data Visualizations

In [171]:
fig = px.scatter_mapbox(indoors, 
                        lat='Lat', lon='Long', 
                        mapbox_style="stamen-terrain", zoom=16.2,
                        color='Floor',
                        size = '4G download rate',
                        color_continuous_scale=px.colors.sequential.Cividis_r,
                        opacity = 0.3,
                        hover_name='Description', 
                        hover_data=['Location ID','4G RSSI signal', '4G upload rate'],
                        title = '4G download rate, By Floors',
 )
fig.update_layout(mapbox_style = 'carto-positron')
fig.show()

In [172]:
fig = px.scatter_mapbox(indoors, 
                        lat='Lat', lon='Long', 
                        mapbox_style="stamen-terrain", zoom=16.2,
                        color='Floor',
                        size = '4G upload rate',
                        color_continuous_scale=px.colors.sequential.Cividis_r,
                        opacity = 0.3,
                        hover_name='Description', 
                        hover_data=['Location ID','4G RSSI signal', '4G upload rate'],
                        title = '4G upload rate, By Floors',
 )
fig.update_layout(mapbox_style = 'carto-positron')
fig.show()

## RSSI Signal Strength by Floor

In [6]:
rssi_max = indoors['4G RSSI signal'].max()
rssi_min = indoors['4G RSSI signal'].min()
range_rssi = [rssi_min, rssi_max] 
print("MAX RSSI: ", rssi_max)
print("MAX RSSI: ", rssi_min)

MAX RSSI:  -71
MAX RSSI:  -200


In [116]:
fig = px.histogram(indoors, '4G RSSI signal', 
                   color = 'Floor', barmode='group',  
                   title='Histogram of 4G RSSI signal, by Revenue',
                  )
#labels={'count':'number of sessions',
#                      'ProductRelated_Duration':'Time Spent Viewing Product-related Webpage',
#                      'Revenue':'Successful Sale',
#                      1:'Success',
#                      0:'No success'
#                     }
fig.show()

#### *Basement floors* RSSI Signal

In [80]:
basement_data = indoors[(indoors['Floor'] == '0')| (indoors['Floor'] == 'A' )| (indoors['Floor'] == 'B')]
fig = px.scatter_mapbox(basement_data, 
                        lat='Lat', lon='Long', 
                        mapbox_style="stamen-terrain", zoom=16.2,
                        color='4G RSSI signal',
                        color_continuous_scale=px.colors.sequential.Cividis_r,
                        range_color = range_rssi,
                        opacity = 1,
                        hover_name='Description', 
                        hover_data=['Location ID','4G RSSI signal'],
                        title = 'RSSI Signal Strength, Basement Floors',
 )
fig.update_layout(mapbox_style = 'carto-positron')
fig.show()

#### *First Floor* RSSI Signal

In [14]:
#GRAPH Basement floors RSSI Signal
first_floor = indoors[indoors['Floor'] == '1']
fig = px.scatter_mapbox(first_floor, 
                        lat='Lat', lon='Long', 
                        mapbox_style="stamen-terrain", zoom=16.2,
                        color='4G RSSI signal',
                        color_continuous_scale=px.colors.sequential.Cividis_r,
                        range_color = range_rssi,
                        opacity = 1,
                        hover_name='Description', 
                        hover_data=['Location ID','4G RSSI signal'],
                        title = 'RSSI Signal Strength, First Floors'
 )
fig.update_layout(mapbox_style = 'carto-positron')
fig.show()

#### *Second Floor* RSSI Signal

In [79]:
second_floor = indoors[indoors['Floor'] == '2']
fig = px.scatter_mapbox(second_floor, 
                        lat='Lat', lon='Long', 
                        mapbox_style="stamen-terrain", zoom=16.2,
                        color='4G RSSI signal',
                        color_continuous_scale=px.colors.sequential.Cividis_r,
                        range_color = range_rssi,
                        opacity = 1,
                        hover_name='Description', 
                        hover_data=['Location ID','4G RSSI signal'],
                        title = 'RSSI Signal Strength, Second Floors',
 )
fig.update_layout(mapbox_style = 'carto-positron')
fig.show()

#### *Third Floor* RSSI Signal

In [17]:
third = indoors[indoors['Floor'] == '3']
fig = px.scatter_mapbox(third, 
                        lat='Lat', lon='Long', 
                        mapbox_style="stamen-terrain", zoom=16.2,
                        color='4G RSSI signal',
                        color_continuous_scale=px.colors.sequential.Cividis_r,
                        range_color = range_rssi,
                        opacity = 1,
                        hover_name='Description', 
                        hover_data=['Location ID','4G RSSI signal'],
                        title = 'RSSI Signal Strenth, Third Floors',
 )
fig.update_layout(mapbox_style = 'carto-positron')
fig.show()

#### *Fourth Floor* RSSI Signal

In [21]:
fourth = indoors[indoors['Floor'] == '4']
fig = px.scatter_mapbox(fourth, 
                        lat='Lat', lon='Long', 
                        mapbox_style="stamen-terrain", zoom=18.2,
                        color='4G RSSI signal',
                        color_continuous_scale=px.colors.sequential.Cividis_r,
                        range_color = range_rssi,
                        opacity = 1,
                        hover_name='Description', 
                        hover_data=['Location ID','4G RSSI signal'],
                        title = 'RSSI Signal Strength, Fourth Floors',
 )
fig.update_layout(mapbox_style = 'carto-positron')
fig.show()

In [173]:
fifth = indoors[indoors['Floor'] == '5']
fig = px.scatter_mapbox(fifth, 
                        lat='Lat', lon='Long', 
                        mapbox_style="stamen-terrain", zoom=18.2,
                        color='4G RSSI signal',
                        color_continuous_scale=px.colors.sequential.Cividis_r,
                        range_color = range_rssi,
                        opacity = 1,
                        hover_name='Description', 
                        hover_data=['Location ID','4G RSSI signal'],
                        title = 'RSSI Signal Strength, Fourth Floors',
 )
fig.update_layout(mapbox_style = 'carto-positron')
fig.show()

## Download Rates by Floor

16.468354430379755

In [120]:
fig = px.histogram(indoors, '4G download rate', 
                   color = 'Floor', 
                   title='Histogram of 4G download rate, by floor',
                   hover_data = ['Description', '4G RSSI signal']
                  )
fig.update_traces(xbins=dict( # bins used for histogram
        start=0.0,
        end=90.0,
        size=5
    ))
fig.show()

In [88]:
download_max = indoors['4G download rate'].max()
download_min = indoors['4G download rate'].min()
range_down = [download_min, download_max] 
print("MAX Upload Rate: ", download_max)
print("MAX Upload Rate: ", download_min)

MAX Upload Rate:  80.3
MAX Upload Rate:  0.0


In [90]:
basement_data = indoors[(indoors['Floor'] == '0')| (indoors['Floor'] == 'A' )| (indoors['Floor'] == 'B')]
fig = px.scatter_mapbox(basement_data, 
                        lat='Lat', lon='Long', 
                        mapbox_style="stamen-terrain", zoom=16.2,
                        size = '4G download rate',
                        color= 'Description',
                        opacity = 0.5,
                        hover_name='Description', 
                        hover_data=['Location ID','4G download rate', '4G upload rate'],
                        title = '4G Download Rate - indoors, Basement Floors',
 )
fig.update_layout(mapbox_style = 'carto-positron')
fig.show()

In [33]:
first = indoors[indoors['Floor'] == '1']
fig = px.scatter_mapbox(first, 
                        lat='Lat', lon='Long', 
                        mapbox_style="stamen-terrain", zoom=16.0,
                        size='4G download rate',
                        color = 'Description',
                        opacity = 1,
                        hover_name='Description', 
                        hover_data=['Location ID','4G download rate', '4G upload rate'],
                        title = '4G Download Rate - indoors, First Floors',
 )
fig.update_layout(mapbox_style = 'carto-positron')
fig.show()

In [38]:
second = indoors[indoors['Floor'] == '2']
fig = px.scatter_mapbox(second, 
                        lat='Lat', lon='Long', 
                        mapbox_style="stamen-terrain", zoom=16.0,
                        color = 'Description',
                        size='4G download rate',
                        opacity = 0.7,
                        hover_name='Description', 
                        hover_data=['Location ID','4G download rate', '4G upload rate'],
                        title = '4G Download Rate - indoors, Second Floors',
 )
fig.update_layout(mapbox_style = 'carto-positron')
fig.show()

In [78]:
third = indoors[indoors['Floor'] == '3']
fig = px.scatter_mapbox(third, 
                        lat='Lat', lon='Long', 
                        mapbox_style="stamen-terrain", zoom=16.0,
                        color = 'Description',
                        size='4G download rate',
                        opacity = 0.7,
                        hover_name='Description', 
                        hover_data=['Location ID','4G download rate', '4G upload rate'],
                        title = '4G Download Rate - indoors, Second Floors',
 )
fig.update_layout(mapbox_style = 'carto-positron')
fig.show()

In [43]:
fourth = indoors[indoors['Floor'] == '4']
fig = px.scatter_mapbox(fourth, 
                        lat='Lat', lon='Long', 
                        mapbox_style="stamen-terrain", zoom=18.0,
                        color = 'Description',
                        size='4G download rate',
                        opacity = 1,
                        hover_name='Description', 
                        hover_data=['Location ID','4G download rate', '4G upload rate'],
                        title = '4G Download Rate - indoors, Fourth Floors',
 )
fig.update_layout(mapbox_style = 'carto-positron')
fig.show()

In [45]:
fifth = indoors[indoors['Floor'] == '5']
fig = px.scatter_mapbox(fifth, 
                        lat='Lat', lon='Long', 
                        mapbox_style="stamen-terrain", zoom=18.0,
                        color = 'Description',
                        size='4G download rate',
                        opacity = 1,
                        hover_name='Description', 
                        hover_data=['Location ID','4G download rate', '4G upload rate'],
                        title = '4G Download Rate - indoors, Fifth Floors',
 )
fig.update_layout(mapbox_style = 'carto-positron')
fig.show()

## Obsevation:


In [84]:
fig = px.histogram(indoors, '4G upload rate', 
                   color = 'Floor', 
                   title='Histogram of 4G upload rate, by floor',
                   hover_data = ['Description', '4G RSSI signal']
                  )
fig.update_traces(xbins=dict( # bins used for histogram
        start=0.0,
        size=2
    ))
#labels={'count':'number of sessions',
#                      'ProductRelated_Duration':'Time Spent Viewing Product-related Webpage',
#                      'Revenue':'Successful Sale',
#                      1:'Success',
#                      0:'No success'
#                     }
fig.show()

In [104]:
upload_max = indoors['4G upload rate'].max()
upload_min = indoors['4G upload rate'].min()
range_up = [upload_min, upload_max] 
print("MAX Upload Rate: ", upload_max)
print("MAX Upload Rate: ", upload_min)

MAX Upload Rate:  33.5
MAX Upload Rate:  0.0


In [66]:
basement_data = indoors[(indoors['Floor'] == '0')| (indoors['Floor'] == 'A' )| (indoors['Floor'] == 'B')]
fig = px.scatter_mapbox(basement_data, 
                        lat='Lat', lon='Long', 
                        mapbox_style="stamen-terrain", zoom=16.2,
                        size = '4G upload rate',
                        color= 'Description',
                        opacity = 0.5,
                        hover_name='Description', 
                        hover_data=['Location ID','4G download rate', '4G upload rate'],
                        title = '4G Download Rate - indoors, Basement Floors',
 )
fig.update_layout(mapbox_style = 'carto-positron')
fig.show()

In [74]:
fifth = indoors[indoors['Floor'] == '1']
fig = px.scatter_mapbox(fifth, 
                        lat='Lat', lon='Long', 
                        mapbox_style="stamen-terrain", zoom=16.0,
                        color = 'Description',
                        size='4G upload rate',
                        opacity = 0.7,
                        hover_name='Description', 
                        hover_data=['Location ID','4G download rate', '4G upload rate'],
                        title = '4G Upload Rate - indoors, First Floors',
 )
fig.update_layout(mapbox_style = 'carto-positron')
fig.show()

In [72]:
fifth = indoors[indoors['Floor'] == '2']
fig = px.scatter_mapbox(fifth, 
                        lat='Lat', lon='Long', 
                        mapbox_style="stamen-terrain", zoom=16.0,
                        color = 'Description',
                        size='4G upload rate',
                        opacity = 0.7,
                        hover_name='Description', 
                        hover_data=['Location ID','4G download rate', '4G upload rate'],
                        title = '4G Upload Rate - indoors, Second Floors',
 )
fig.update_layout(mapbox_style = 'carto-positron')
fig.show()

In [71]:
fifth = indoors[indoors['Floor'] == '3']
fig = px.scatter_mapbox(fifth, 
                        lat='Lat', lon='Long', 
                        mapbox_style="stamen-terrain", zoom=16.0,
                        color = 'Description',
                        size='4G upload rate',
                        opacity = 0.7,
                        hover_name='Description', 
                        hover_data=['Location ID','4G download rate', '4G upload rate'],
                        title = '4G Upload Rate - indoors, Third Floors',
 )
fig.update_layout(mapbox_style = 'carto-positron')
fig.show()

In [67]:
fifth = indoors[indoors['Floor'] == '4']
fig = px.scatter_mapbox(fifth, 
                        lat='Lat', lon='Long', 
                        mapbox_style="stamen-terrain", zoom=18.0,
                        color = 'Description',
                        size='4G upload rate',
                        opacity = 1,
                        hover_name='Description', 
                        hover_data=['Location ID','4G download rate', '4G upload rate'],
                        title = '4G Upload Rate - indoors, Fourth Floors',
 )
fig.update_layout(mapbox_style = 'carto-positron')
fig.show()

In [64]:
fifth = indoors[indoors['Floor'] == '5']
fig = px.scatter_mapbox(fifth, 
                        lat='Lat', lon='Long', 
                        mapbox_style="stamen-terrain", zoom=18.0,
                        color = 'Description',
                        size='4G upload rate',
                        opacity = 1,
                        hover_name='Description', 
                        hover_data=['Location ID','4G download rate', '4G upload rate'],
                        title = '4G Download Rate - indoors, Fifth Floors',
 )
fig.update_layout(mapbox_style = 'carto-positron')
fig.show()

### Outdoor Area

In [122]:
fig = px.histogram(outdoors, '4G download rate', 
                   color = 'Description', 
                   title='Histogram of 4G download rate, by floor',
                   hover_data = ['Description', '4G RSSI signal']
                  )
fig.update_traces(xbins=dict( # bins used for histogram
        start=0.0,
        size=2
    ))

In [112]:
fig = px.histogram(outdoors, '4G upload rate', 
                   color = 'Description', 
                   title='Histogram of 4G upload rate, by floor',
                   hover_data = ['Description', '4G download rate']
                  )
fig.update_traces(xbins=dict( # bins used for histogram
        start=0.0,
        size=2
    ))

In [115]:
fig = px.histogram(outdoors, '4G RSSI signal', 
                   color = 'Description', 
                   title='Histogram of 4G RSSI signal, by floor',
                   hover_data = ['Description', '4G download rate', '4G upload rate']
                  )
fig.update_traces(xbins=dict( # bins used for histogram
        size=2
    ))

In [125]:
fig = px.scatter_mapbox(outdoors, 
                        lat='Lat', lon='Long', 
                        mapbox_style="stamen-terrain", zoom=16.0,
                        color = 'Description',
                        size='4G upload rate',
                        opacity = 1,
                        hover_name='Description', 
                        hover_data=['Location ID','4G download rate', '4G upload rate'],
                        title = '4G Upload Rate - Outdoors',
 )
fig.update_layout(mapbox_style = 'carto-positron')
fig.show()

In [127]:
fig = px.scatter_mapbox(outdoors, 
                        lat='Lat', lon='Long', 
                        mapbox_style="stamen-terrain", zoom=16.0,
                        color = 'Description',
                        size='4G download rate',
                        opacity = 1,
                        hover_name='Description', 
                        hover_data=['Location ID','4G download rate', '4G upload rate'],
                        title = '4G Download Rate - Outdoors',
 )
fig.update_layout(mapbox_style = 'carto-positron')
fig.show()

In [129]:
fig = px.scatter_mapbox(outdoors, 
                        lat='Lat', lon='Long', 
                        mapbox_style="stamen-terrain", zoom=16.0,
                        color='4G RSSI signal',
                        opacity = 1,
                        hover_name='Description', 
                        hover_data=['Location ID','4G download rate', '4G upload rate'],
                        title = '4G RSSI Signal - Outdoors',
 )
fig.update_layout(mapbox_style = 'carto-positron')
fig.show()

### Indoor Areas Stats

In [180]:
print("Stats for INDOORS download rate")
print("mean", indoors['4G download rate'].mean())
print("standard deviation", indoors['4G download rate'].std())
print("range", range_down)
print("------------")
print("Stats for INDOORS upload rate")
print("mean", indoors['4G upload rate'].mean())
print("standard deviation", indoors['4G upload rate'].std())
print("range", range_up)
print("------------")
print("Stats for INDOORS RSSI Signal")
print("mean", indoors['4G download rate'].mean())
print("standard deviation", indoors['4G RSSI signal'].std())
print("range*", range_rssi)

print("*note: the -200 in rssi range reprents no signal, if we use -infinity it would skew the color map too much")

Stats for INDOORS download rate
mean 16.468354430379755
standard deviation 27.390081230647827
range [0.0, 80.3]
------------
Stats for INDOORS upload rate
mean 7.994936708860766
standard deviation 7.982045537171615
range [0.0, 33.5]
------------
Stats for INDOORS RSSI Signal
mean 16.468354430379755
standard deviation 25.651365166229834
range* [-200, -71]
*note: the -200 in rssi range reprents no signal, if we use -infinity it would skew the color map too much


In [186]:
print("Stats for OUTDOORS download rate")
print("mean", outdoors['4G download rate'].mean())
print("standard deviation", outdoors['4G download rate'].std())
range_out_max = outdoors['4G download rate'].max()
range_out_min = outdoors['4G download rate'].min()
range_out_down = [range_out_min, range_out_max]
print("range", range_out_down)

print("------------")

print("Stats for indoors upload rate")
print("mean", outdoors['4G upload rate'].mean())
range_out_max = outdoors['4G upload rate'].max()
range_out_min = outdoors['4G upload rate'].min()
range_out_up = [range_out_min, range_out_max]
print("standard deviation", outdoors['4G upload rate'].std())
print("range", range_out_up)
print("------------")
print("Stats for indoors RSSI Signal")
print("mean", outdoors['4G download rate'].mean())
print("standard deviation", outdoors['4G RSSI signal'].std())
print("range*", range_rssi)

print("*note: the -200 in rssi range reprents no signal, if we use -infinity it would skew the color map too much")

Stats for OUTDOORS download rate
mean 8.254545454545456
standard deviation 7.633617832464578
range [0.8, 33.9]
------------
Stats for indoors upload rate
mean 5.5590909090909095
standard deviation 1.3164378671280532
range [4.6, 8.4]
------------
Stats for indoors RSSI Signal
mean 8.254545454545456
standard deviation 6.129906816237713
range* [-200, -71]
*note: the -200 in rssi range reprents no signal, if we use -infinity it would skew the color map too much


### Find the low download, upload rate areas:

In [151]:
mean_indoors_down = indoors['4G download rate'].mean()
fig = px.bar(indoors.sort_values(by = '4G download rate', ascending = True), x = 'Description', y = '4G download rate', 
       color = 'Floor', color_discrete_sequence= px.colors.qualitative.Vivid, barmode='group',
       title = 'Indoors Download Rate by Location',
       hover_data = ['4G RSSI signal', '4G upload rate'],
       labels= {'Description': "Location"})
fig.add_hline(indoors['4G download rate'].mean(), line_dash = 'dash')

In [152]:
mean_indoors_up = indoors['4G upload rate'].mean()
fig = px.bar(indoors.sort_values(by = '4G upload rate', ascending = True), x = 'Description', y = '4G download rate', 
       barmode='group',
       color = 'Floor', color_discrete_sequence= px.colors.qualitative.Vivid,
       title = 'Indoors Upload Rate by Location',
       hover_data = ['4G RSSI signal', '4G upload rate'],
       labels= {'Description': "Location"})
fig.add_hline(mean_indoors_up, line_dash = 'dash')

In [150]:
mean_indoors_rssi = indoors['4G RSSI signal'].mean()
fig = px.bar(indoors.sort_values(by = '4G RSSI signal', ascending = True), x = 'Description', y = '4G RSSI signal', 
       color = 'Floor', color_discrete_sequence= px.colors.qualitative.Vivid, barmode = 'group',
       title = 'Indoors RSSI Signal by Location',
       hover_data = ['4G RSSI signal', '4G upload rate'],
       labels= {'Description': "Location"})
fig.add_hline(mean_indoors_rssi, line_dash = 'dash')

## Outdoors Statistics

In [164]:
mean_outdoors_down = outdoors['4G download rate'].mean()
fig = px.bar(outdoors.sort_values(by = '4G download rate', ascending = True), x = 'Description', y = '4G download rate', 
       title = 'Outdoors Download Rate by Location',
       hover_data = ['4G RSSI signal', '4G upload rate'],
       labels= {'Description': "Location"})
fig.add_hline(mean_outdoors_down, line_dash = 'dash', annotation_text="outdoors download rate avg")
fig.add_hline(mean_indoors_down, line_dash = 'dash', line_color= 'red',annotation_text="indoors download rate avg")
fig.update_annotations(align='left')

In [177]:
mean_outdoors_up = outdoors['4G upload rate'].mean()
fig = px.bar(outdoors.sort_values(by = '4G upload rate', ascending = True), x = 'Description', y = '4G upload rate', 
       
       title = 'Outdoors Upload Rate by Location',
       hover_data = ['4G download rate', '4G download rate'],
       labels= {'Description': "Location"})
fig.add_hline(mean_indoors_up, line_dash = 'dash', line_color = 'red, annotation_text="indoors upload rate avg")
fig.add_hline(mean_outdoors_up, line_dash = 'dash', annotation_text="outdoors download rate avg")

SyntaxError: EOL while scanning string literal (3313682613.py, line 7)

In [179]:
mean_outdoors_rssi = outdoors['4G RSSI signal'].mean()
fig = px.bar(outdoors.sort_values(by = '4G RSSI signal', ascending = True), x = 'Description', y = '4G RSSI signal', 
       
       title = 'Outdoors RSSI Signal by Location',
       hover_data = ['4G download rate', '4G upload rate'],
       labels= {'Description': "Location"})
fig.add_hline(mean_indoors_rssi, line_dash = 'dash',line_color = 'red',  annotation_text="indoors RSSI avg")
fig.add_hline(mean_outdoors_rssi, line_dash = 'dash', annotation_text="outdoors RSSI avg")

In [ ]:
indoors[['']]

### Conclusion:
4G download rates are generally not very good, with a range of [0, 80.3] but a mean of 